// ng-gallery/esm2022/lib/directives/gallery-thumb-def.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
class GalleryThumbDef {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    // Make sure the template checker knows the type of the context with which the
    // template of this directive will be rendered
    static ngTemplateContextGuard(directive, context) {
        return true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: GalleryThumbDef, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.1", type: GalleryThumbDef, isStandalone: true, selector: "[galleryThumbDef]", ngImport: i0 }); }
}
export { GalleryThumbDef };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: GalleryThumbDef, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[galleryThumbDef]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
