// ng-gallery/esm2022/lightbox/lightbox.directive.mjs
import { Directive, Input } from '@angular/core';
import { fromEvent, tap, Subscription } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./lightbox.service";
class LightboxDirective {
    constructor(_lightbox, _el, _renderer) {
        this._lightbox = _lightbox;
        this._el = _el;
        this._renderer = _renderer;
        this.clickEvent = Subscription.EMPTY;
        this.index = 0;
        this.id = 'root';
    }
    ngOnInit() {
        this._renderer.setStyle(this._el.nativeElement, 'cursor', 'pointer');
        this.clickEvent = fromEvent(this._el.nativeElement, 'click').pipe(tap(() => this._lightbox.open(this.index, this.id))).subscribe();
    }
    ngOnDestroy() {
        this.clickEvent.unsubscribe();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: LightboxDirective, deps: [{ token: i1.Lightbox }, { token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.1", type: LightboxDirective, isStandalone: true, selector: "[lightbox]", inputs: { index: ["lightbox", "index"], id: ["gallery", "id"] }, ngImport: i0 }); }
}
export { LightboxDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.1", ngImport: i0, type: LightboxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[lightbox]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.Lightbox }, { type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { index: [{
                type: Input,
                args: ['lightbox']
            }], id: [{
                type: Input,
                args: ['gallery']
            }] } });
